### PR TITLE
Add ACFWindowActive to /redfish/v1

### PR DESCRIPTION
**Pull request by joseph-reynolds**
Partial fix for SW546530

This adds the Oem.IBM.ACFWindowActive boolean property to the Redfish service root at URI /redfish/v1.

Tested:
Via curl https://${BMC}/redfish/v1
Tested values true, false, and panel app not running.
Simulated pressing function 74 via busctl set-property "com.ibm.PanelApp"
  "/com/ibm/panel_app" "com.ibm.panel" "ACFWindowActive" b true

Signed-off-by: Joseph Reynolds <joseph-reynolds@charter.net>

### CODE REVIEW

File: redfish-core/lib/service_root.hpp [hunk 1]
```diff
@@ -123,6 +123,34 @@ inline void
...
+                BMCWEB_LOG_ERROR
+                    << "Failed to read ACFWindowActive property";
+            }
+            else
+            {
+                const bool* isACFWindowActive = std::get_if<bool>(&retVal);
+                if (isACFWindowActive == nullptr)
+                {
+                    BMCWEB_LOG_ERROR << "nullptr for ACFWindowActive";
+                    messages::internalError(asyncResp->res);
```
Comment (gtmills) [inline redfish-core/lib/service_root.hpp:141]: more typical in bmcweb is 
            if (ec)
            {
                BMCWEB_LOG_ERROR
                    << "Failed to read ACFWindowActive property";
                    
                    return;
            }
            
                const bool* isACFWindowActive = std::get_if<bool>(&retVal);
                if (isACFWindowActive == nullptr)
                {
                    BMCWEB_LOG_ERROR << "nullptr for ACFWindowActive";
                    messages::internalError(asyncResp->res);
                    return;
                 }

File: redfish-core/lib/service_root.hpp [hunk 2]
```diff
@@ -123,6 +123,34 @@ inline void
...
         redfishDateTimeOffset.second;
 
+    // Report if ACFWindowActive
+    crow::connections::systemBus->async_method_call(
+        [asyncResp](const boost::system::error_code ec,
+                    const std::variant<bool>& retVal) {
+            if (ec)
+            {
+                BMCWEB_LOG_ERROR
+                    << "Failed to read ACFWindowActive property";
```
Comment (gtmills) [inline redfish-core/lib/service_root.hpp:133]: Your call if you want an internal error, it more common to have one, but I guess someone could argue this isn't an important property? What would the GUI do if this isn't there?

Comment (joseph-reynolds) [reply redfish-core/lib/service_root.hpp:133]: Intended usage is: if the ACFWindowActive property is present and has the true value, then show the "Upload ACF" button on the login page (before authentication and before a session is created).

File: redfish-core/lib/service_root.hpp (hunk 1 above)
Comment (joseph-reynolds) [reply redfish-core/lib/service_root.hpp:141]: Thanks.   I'll break out handleACFWindowActive as a separate function so I can structure the code like this.

File: redfish-core/lib/service_root.hpp (hunk 2 above)
Comment (joseph-reynolds) [reply redfish-core/lib/service_root.hpp:133]: I'll change this to default value false.

File: static/redfish/v1/JsonSchemas/OemServiceRoot/index.json [hunk 3]
```diff
@@ -85,6 +85,15 @@
                         "string",
                         "null"
                     ]
+                },
```
[the remark this answers is not in the document]
Comment (joseph-reynolds) [reply static/redfish/v1/JsonSchemas/OemServiceRoot/index.json:88]: Yes, I see it now.  Thanks, I'll add it.

File: redfish-core/lib/service_root.hpp [hunk 4]
```diff
@@ -23,6 +23,40 @@
...
+                // Default value when panel app is unreachable.
+                asyncResp->res.jsonValue["Oem"]["IBM"]["ACFWindowActive"] =
+                    false;
```
Comment (gtmills) [inline redfish-core/lib/service_root.hpp:39]: return here?

File: redfish-core/lib/service_root.hpp [hunk 5]
```diff
@@ -23,6 +23,40 @@
...
+            }
+            const bool* isACFWindowActive = std::get_if<bool>(&retVal);
+            if (isACFWindowActive == nullptr)
+            {
+                BMCWEB_LOG_ERROR << "nullptr for ACFWindowActive";
+                messages::internalError(asyncResp->res);
+            }
```
Comment (gtmills) [inline redfish-core/lib/service_root.hpp:46]: return here and get rid of the else?

File: redfish-core/lib/service_root.hpp (hunk 4 above)
Comment (gtmills) [reply redfish-core/lib/service_root.hpp:39]: e.g. from line 69:

                subtree) {
            if (ec)
            {
                BMCWEB_LOG_DEBUG << "DBUS response error";
                messages::internalError(asyncResp->res);
                return;
            }

Comment (joseph-reynolds) [reply redfish-core/lib/service_root.hpp:39]: Good catch.  I'll return here.

File: redfish-core/lib/service_root.hpp (hunk 5 above)
Comment (joseph-reynolds) [reply redfish-core/lib/service_root.hpp:46]: Yes, I'll return here to follow the coding style.

**Comment by gtmills** [comment]
@joseph-reynolds Validator passed?

**Review by gtmills — APPROVED** [review]

**Comment by joseph-reynolds** [comment]
Partially fixes SW546530.

**Comment by joseph-reynolds** [comment]
Redfish Validator failures:
```
*** /redfish/v1/Systems/system/LogServices/EventLog/Entries/1481
	 Type (LogEntry.v1_9_0.LogEntry), GET SUCCESS (time: 0)
OpenBMC: EntityType resource does not contain required @odata.id property, attempting default /OpenBMC
	  FAIL...
...
*** /redfish/v1/Systems/system/LogServices/CELog/Entries/1472
	 Type (LogEntry.v1_9_0.LogEntry), GET SUCCESS (time: 0)
OpenBMC: EntityType resource does not contain required @odata.id property, attempting default /OpenBMC
	  FAIL...
...et.seq...
*** /redfish/v1/SessionService/Sessions/Vi1i6RUoC5
	 Type (Session.v1_3_0.Session), GET SUCCESS (time: 0:00:00.014431)
OpenBMC: EntityType resource does not contain required @odata.id property, attempting default /OpenBMC
	  FAIL...

...et.seq...

1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/EventLog/Entries/1481
1 failProp errors in /redfish/v1/Systems/system/LogServices/EventLog/Entries/1481
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1472
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1472
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1473
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1473
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1474
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1474
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1475
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1475
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1476
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1476
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1477
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1477
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1478
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1478
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1479
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1479
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1480
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1480
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1482
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1482
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1483
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1483
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1484
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1484
1 err.OemLogEntry.v1_0_0.LogEntry errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1485
1 failProp errors in /redfish/v1/Systems/system/LogServices/CELog/Entries/1485
1 err.OemSession.v1_0_0.Session errors in /redfish/v1/SessionService/Sessions/Vi1i6RUoC5
1 failProp errors in /redfish/v1/SessionService/Sessions/Vi1i6RUoC5
1 err.OemSession.v1_0_0.Session errors in /redfish/v1/SessionService/Sessions/CU8f0Qk3N2
1 failProp errors in /redfish/v1/SessionService/Sessions/CU8f0Qk3N2
1 err.OemSession.v1_0_0.Session errors in /redfish/v1/SessionService/Sessions/rV3obRkQgS
1 failProp errors in /redfish/v1/SessionService/Sessions/rV3obRkQgS
err.OemLogEntry.v1_0_0.LogEntry: 14   
err.OemSession.v1_0_0.Session: 3   
failProp: 17   
invalidPropertyValue: 60   
metadataNamespaces: 3289   
pass: 12344   
passAction: 19   
passGet: 819   
passRedfishUri: 770   
repeat: 6   
skipNoSchema: 3   
skipOptional: 19824   
unverifiedAdditional.complex: 8   
warnDeprecated: 5   
warnTrailingSlashLink: 6   
warningPresent: 56   
Validation has failed: 17 problems found
```

All failures were not related to the /redfish/v1 Service root which this PR changes.

**Comment by joseph-reynolds** [comment]
Merged by https://github.ibm.com/openbmc/openbmc/pull/2392

**Pull request merged**